### PR TITLE
Allow privateips on console proxy

### PR DESCRIPTION
**Pull request by GaOrtiga**
### Description

Currently the configuration `consoleproxy.url.domain` only accepts domain names as a valid value; however, some users have reported the desire to use private IPs as a value for this configuration. This is useful for CI implementation and local labs.

The main goal of this PR is to allow operators to use IPs as valid configurations as well.

The validations for the configuration values have also been refactored into new methods to facilitate unity tests and clean up the code.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### CODE REVIEW

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -1098,35 +1099,36 @@ private String validateConfigurationValue(final String name, String value, final
...
-        final Config c = Config.getConfig(name);
-        if (c == null) {
-            s_logger.warn("Did not find configuration " + name + " in Config.java. Perhaps moved to ConfigDepot");
+        final Config configuration = Config.getConfig(name);
+        if (configuration == null) {
+            s_logger.warn(String.format("Configuration [%s] could not be found in Config.java. Perhaps moved to ConfigDepot.", name));
             final ConfigKey<?> configKey = _configDepot.get(name);
             if(configKey == null) {
-                s_logger.warn("Did not find configuration " + name + " in ConfigDepot too.");
+                s_logger.warn(String.format("Configuration [%s] could not be found in ConfigDepot."));
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1113]:
```suggestion
                s_logger.warn(String.format("Configuration [%s] could not be found in ConfigDepot.", configKey));
```

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -1171,147 +1173,205 @@ private String validateConfigurationValue(final String name, String value, final
...
+                //The value needs to be between 0 to 255 because the mac generation needs a value of 8 bit
                 //0 value is considered as disable.
                 if(val < 0 || val > 255){
-                    throw new InvalidParameterValueException(name+" value should be between 0 and 255. 0 value will disable this feature");
+                    throw new InvalidParameterValueException(String.format("[%s] value should be between 0 and 255. 0 value will disable this feature.", name));
                 }
             } catch (final NumberFormatException e) {
-                s_logger.error("There was an error trying to parse the integer value for:" + name);
-                throw new InvalidParameterValueException("There was an error trying to parse the integer value for:" + name);
+                s_logger.error(String.format("There was an error trying to parse the integer value for: [%s].", name));
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1182]: let's extract this format string in a constant. it is used in four places.
Also let's construct the message one time for both the log and the exception below.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -1149,20 +1151,20 @@ private String validateConfigurationValue(final String name, String value, final
...
-            if (overprovisioningFactorsForValidation.contains(name) && Float.parseFloat(value) <= 0f) {
-                final String msg = name + " should be greater than 0";
+            if (overprovisioningFactorsForValidation.contains(name) && Float.parseFloat(value) < 1f) {
+                final String msg = String.format("[%s] should be greater than or equal to 1.", name);
                 s_logger.error(msg);
                 throw new InvalidParameterValueException(msg);
             }
         } catch (final NumberFormatException e) {
-            final String msg = "There was an error trying to parse the float value for: " + name;
+            final String msg = String.format("There was an error trying to parse the float value for: [%s].", name);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1160]: this format string is used at 1222 and 1223 as well. let's define a constant?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -1171,147 +1173,205 @@ private String validateConfigurationValue(final String name, String value, final
...
+        }
+        s_logger.error(String.format("Netmask [%1$s] is not a valid netmask for configuration variable [%2$s].", value, name));
+        return "Please enter a valid netmask.";
+    }
+
+    protected String validateRangeHypervisorList(String value) {
+        final String[] hypervisors = value.split(",");
+        for (final String hypervisor : hypervisors) {
+            if (HypervisorType.getType(hypervisor) == HypervisorType.Any || HypervisorType.getType(hypervisor) == HypervisorType.None) {
+                return String.format("Please enter a valid hypervisor type.");
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1325]: why the `String.format()` here? did you intent to add values (like `hypervisor` or `HypervisorType`)?

**Review by DaanHoogland — COMMENTED** [review]
small issues with the code so far.

**Comment by DaanHoogland** [comment]
@GaOrtiga dio you expect to have this finished/ready for review in the coming weeks?

**Comment by DaanHoogland** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=6808)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=BUG) [E](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=BUG) [1 Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6808&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6808&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6808&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=CODE_SMELL) [D](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=CODE_SMELL) [270 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6808&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6808&metric=new_coverage&view=list) [34.6% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6808&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6808&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6808&metric=new_duplicated_lines_density&view=list)

@GaOrtiga 270 code smells seems a bit daunting. Are you still working on this?

**Comment by DaanHoogland** [comment]
@GaOrtiga as this is still in draft, should it be pushed to a later milestone? i.e. 4.19?

**Comment by GaOrtiga** [comment]
> @GaOrtiga as this is still in draft, should it be pushed to a later milestone? i.e. 4.19?

Sorry about the delay to finish this, while creating the tests I found the error that was fixed here: https://github.com/apache/cloudstack/pull/6910
This error needed to be fixed for the tests to run, so now that the fix was merged I came back for this PR.
I believe I have finished the changes, it is working as expected so far.
I will run some more tests and hopefully take out of draft this week.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/test/java/com/cloud/configuration/ConfigurationManagerImplTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
+package com.cloud.configuration;
```
Comment (DaanHoogland) [inline server/src/test/java/com/cloud/configuration/ConfigurationManagerImplTest.java:1]: license needed

**Review by DaanHoogland — CHANGES_REQUESTED** [review]
clgtm, but a license header is needed/missing

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

File: server/src/main/java/com/cloud/configuration/Config.java [hunk 6]
```diff
@@ -426,7 +426,7 @@ public enum Config {
             "Console proxy command port that is used to communicate with management server",
             null),
     ConsoleProxyRestart("Console Proxy", AgentManager.class, Boolean.class, "consoleproxy.restart", "true", "Console proxy restart flag, defaulted to true", null),
-    ConsoleProxyUrlDomain("Console Proxy", AgentManager.class, String.class, "consoleproxy.url.domain", "", "Console proxy url domain", "domainName"),
+    ConsoleProxyUrlDomain("Console Proxy", AgentManager.class, String.class, "consoleproxy.url.domain", "", "Console proxy url domain", "domainName","privateip"),
```
Comment (RodrigoDLopez) [inline server/src/main/java/com/cloud/configuration/Config.java:429]:
```suggestion
    ConsoleProxyUrlDomain("Console Proxy", AgentManager.class, String.class, "consoleproxy.url.domain", "", "Console proxy url domain", "domainName", "privateip"),
```

**Review by JoaoJandre — APPROVED** [review]
LGTM

**Review by RodrigoDLopez — APPROVED** [review]
LGTM. 
Tested locally and everything work like a charm

**Comment by DaanHoogland** [comment]
@GaOrtiga,
Do you plan to document this? (in https://github.com/apache/cloudstack-documentation)?
And do you plan to apply @RodrigoDLopez 'extra space suggestion?

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 7]
```diff
@@ -1222,96 +1222,145 @@ private String validateConfigurationValue(final String name, String value, final
...
+        final int val = Integer.parseInt(value);
+        if (val < min || val > max) {
+            s_logger.error(String.format("Invalid value for configuration [%s]. Please enter a value in the range [%s].", name, range));
+            return String.format("The provided value is not valid for this configuration. Please enter an integer in the range: [%s]", range);
+        }
+        return null;
+    }
+
+    /**
+     * This method checks if the value for the configuration is valid for any of the ranges selected.
```
Comment (stephankruggg) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1260]:
```suggestion
     * Checks if the value for the configuration is valid for any of the ranges selected.
```

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 8]
```diff
@@ -1222,96 +1222,145 @@ private String validateConfigurationValue(final String name, String value, final
...
     }
 
+    /**
+     * Valid values are instance names, the only restriction is that they may not have hyphens, spaces or plus signs.
+     */
+    protected String validateRangeInstanceName(String value) {
+        if (NetUtils.verifyInstanceName(value)) {
+            return null;
+        }
+        return "a valid instance name (instance names cannot contain hyphen, space or plus sign)";
```
Comment (stephankruggg) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:1330]:
```suggestion
        return "a valid instance name (instance names cannot contain hyphens, spaces or plus signs)";
```

**Review by stephankruggg — APPROVED** [review]

**Comment by GaOrtiga** [comment]
@DaanHoogland 

`Do you plan to document this? `
Initially I did not plan on documenting since it is not a major change, but I can do it.

`And do you plan to apply @RodrigoDLopez 'extra space suggestion?`
I applied the suggestion, is there a concern regarding it?

**Comment by DaanHoogland** [comment]
> @DaanHoogland
> 
> `Do you plan to document this? ` Initially I did not plan on documenting since it is not a major change, but I can do it.
> 
> `And do you plan to apply @RodrigoDLopez 'extra space suggestion?` I applied the suggestion, is there a concern regarding it?

No concerns, I just wodered what to test/look for. I'm fine with merging as is but I think it will not be for those that do not know beforehand what to expect, so if you an find any documentation on the console proxy domain configuration it would be nice if you can add it.

**Comment by GaOrtiga** [comment]
@DaanHoogland 
I applied one of the suggestions but forgot to update the tests with the current text. Now this test is broken, so, should we revert it or create a fix?

**Comment by weizhouapache** [comment]
> @DaanHoogland I applied one of the suggestions but forgot to update the tests with the current text. Now this test is broken, so, should we revert it or create a fix?

@GaOrtiga 
I just faced the same issue.
please create a pull request to fix the test asap

**Comment by GaOrtiga** [comment]
I created the PR https://github.com/apache/cloudstack/pull/7018
Sorry for the mistake.

**Comment by weizhouapache** [comment]
no worries @GaOrtiga 
pr has been mergd.

**Pull request merged**